File: src/nl/wernerkroneman/Drawy/GlRenderer/Renderer.java
package nl.wernerkroneman.Drawy.GlRenderer;

import com.jogamp.opengl.GL3;
import com.jogamp.opengl.util.glsl.ShaderCode;
import com.jogamp.opengl.util.glsl.ShaderProgram;
import nl.wernerkroneman.Drawy.ConcreteModelling.Drawable;
import nl.wernerkroneman.Drawy.ConcreteModelling.Scene;
import nl.wernerkroneman.Drawy.ConcreteModelling.SceneNode;
import org.joml.Matrix4d;
import org.joml.MatrixStackd;

import static com.jogamp.opengl.GL.*;
import static com.jogamp.opengl.GL2ES2.GL_FRAGMENT_SHADER;
import static com.jogamp.opengl.GL2ES2.GL_VERTEX_SHADER;

/**
 * Created by werner on 29-12-16.
 */
public class Renderer {
    static final String SHADERS_ROOT = "shaders";  // Renderer
    private int shaderProgram;

    private MatrixStackd matrixStack = new MatrixStackd(100);
    private Matrix4d view = new Matrix4d();

    /**
     * Render the shape (triangle)
     *
     * @param scene The scene that we're trying to draw.
     * @param gl    The GL3 context provided by JOML
     */
    void render(Scene scene, GL3 gl) {

        // clear the framebuffer
        gl.glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);

        gl.glUseProgram(shaderProgram);

        // Push a matrix so the transformations in this frame do not affect those in the next
        matrixStack.pushMatrix();

        // Compute the view matrix
        view.identity().lookAt(Scene.EYE, Scene.LOOKAT_CENTER, Scene.UP);

        // Apply camera transform
        matrixStack.mul(view);

        // Scale the y by -1 to get a left-handed coordinate system
        // (would it be better to stick to right-handed, and put the translation in TransitionState?)
        matrixStack.scale(1, -1, 1);

        drawSceneNode(scene.getRootSceneNode(), gl.getGL3());

        matrixStack.popMatrix();

    }

    void initShaders(GL3 gl3) {
        ShaderCode fragShader = ShaderCode.create(gl3, GL_FRAGMENT_SHADER, this.getClass(), SHADERS_ROOT, null,
                "default_fragment", "frag", null, false);
        ShaderCode vertShader = ShaderCode.create(gl3, GL_VERTEX_SHADER, this.getClass(), SHADERS_ROOT, null,
                "default_vertex", "vert", null, false);

        ShaderProgram prog = new ShaderProgram();
        //prog.add(vertShader);
        prog.add(fragShader);

        prog.init(gl3);

        shaderProgram = prog.program();

        gl3.glBindAttribLocation(shaderProgram, ShaderPositions.POSITION, "position");
        gl3.glBindAttribLocation(shaderProgram, ShaderPositions.COLOR, "color");

        prog.link(gl3, System.err);

        vertShader.destroy(gl3);
        fragShader.destroy(gl3);
    }

    private void drawSceneNode(SceneNode rootSceneNode, GL3 gl) {
        matrixStack.pushMatrix();

        matrixStack.mul(rootSceneNode.getTransform()); // Apply transform

        for (SceneNode child : rootSceneNode.getChildren()) {
            drawSceneNode(child, gl); // Draw any children recursively
        }

        for (Drawable ent : rootSceneNode.getDrawables()) {
            draw(ent, gl); // Draw all entities in this node
        }

        matrixStack.popMatrix(); // Undo transforms
    }

    private void draw(Drawable ent, GL3 gl3) {

        gl3.glUseProgram(shaderProgram);

        assert ent.getMesh() instanceof GlMesh;

        GlMesh mesh = (GlMesh) ent.getMesh();

        gl3.glBindVertexArray(mesh.getVertexArrayID());

        gl3.glDrawArrays(GL_TRIANGLES, 1, mesh.getVerticesInBuffer());

        gl3.glBindVertexArray(0);
    }
}

File: src/main/java/nl/wernerkroneman/Drawy/GlRenderer/Renderer.kt
/*
 * Copyright (c) 2017 Werner Kroneman
 *
 * This file is part of DrawyMcDrawface.
 *
 * DrawyMcDrawface is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DrawyMcDrawface is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DrawyMcDrawface.  If not, see <http://www.gnu.org/licenses/>.
 */

package nl.wernerkroneman.Drawy.GlRenderer

import com.jogamp.opengl.GL3
import nl.wernerkroneman.Drawy.ConcreteModelling.Drawable
import nl.wernerkroneman.Drawy.ConcreteModelling.Material
import nl.wernerkroneman.Drawy.ConcreteModelling.Scene
import nl.wernerkroneman.Drawy.ConcreteModelling.SceneNode
import org.joml.Matrix4d
import org.joml.MatrixStackd

import com.jogamp.opengl.GL.*

/**
 * Created by werner on 29-12-16.
 */
class Renderer {

    private val matrixStack = MatrixStackd(100)
    private val view = Matrix4d()
    private val proj = Matrix4d()
    private var shaderProgram: ShaderProgram? = null

    /**
     * Render the shape (triangle)

     * @param scene The scene that we're trying to draw.
     * *
     * @param gl    The GL3 context provided by JOML
     */
    fun render(scene: Scene, gl: GL3) {

        gl.glEnable(GL_DEPTH_TEST)

        // clear the framebuffer

        gl.glClear(GL_COLOR_BUFFER_BIT or GL_DEPTH_BUFFER_BIT)

        shaderProgram!!.use(gl)

        // Push a matrix so the transformations in this frame do not affect those in the next
        matrixStack.pushMatrix()

        // Compute the view matrix
        view.identity().lookAt(Scene.EYE, Scene.LOOKAT_CENTER, Scene.UP)

        // Apply camera transform
        matrixStack.mul(view)

        drawSceneNode(scene.rootSceneNode, gl.gL3)

        matrixStack.popMatrix()

    }

    fun initShaders(gl3: GL3) {
        shaderProgram = ShaderProgram("shaders/default_vertex.vert", "shaders/default_fragment.frag", gl3)
    }

    private fun drawSceneNode(sceneNode: SceneNode, gl: GL3) {

        matrixStack.pushMatrix()
        matrixStack.mul(sceneNode.transform) // Apply transform

        for (child in sceneNode.children) {
            drawSceneNode(child, gl) // Draw any children recursively
        }

        for (ent in sceneNode.drawables) {
            draw(ent, gl) // Draw all entities in this node
        }

        matrixStack.popMatrix() // Undo transforms
    }

    private fun draw(ent: Drawable, gl3: GL3) {

        shaderProgram!!.use(gl3)

        useTopMatrix(gl3)

        useMaterial(ent.material, gl3)

        assert(ent.mesh is GlMesh)

        val mesh = ent.mesh as GlMesh

        if (mesh.isDirty) {
            mesh.recomputeVBO(gl3)
        }

        gl3.glBindVertexArray(mesh.vertexArrayID)

        gl3.glDrawArrays(GL_TRIANGLES, 0, mesh.verticesInBuffer)

        gl3.glBindVertexArray(0)
    }

    private fun useMaterial(material: Material, gl: GL3) {

        shaderProgram!!.setUniform("mat_ambient", material.ambient, gl)
        shaderProgram!!.setUniform("mat_diffuse", material.diffuse, gl)
        shaderProgram!!.setUniform("mat_specular", material.specular, gl)
        shaderProgram!!.setUniform("mat_emissive", material.emissive, gl)

    }

    /**
     * Send the current top matrix and projection matrix to OpenGL.
     */
    private fun useTopMatrix(gl: GL3) {

        // Compute modelviewprojection as projection * modelview to get full transformation.
        val mvp = Matrix4d(proj).mul(matrixStack)

        // Set uniforms
        shaderProgram!!.setUniform("modelviewproj", mvp, gl)
        shaderProgram!!.setUniform("modelview", matrixStack, gl)
    }

    /**
     * Recopute the current projection matrix for a given window size.

     * @param width  Window lateral in pixels
     * *
     * @param height Window height in pixels
     */
    fun computeProjection(width: Int, height: Int) {
        // Compute projection matrix
        proj.identity().perspective(Math.toRadians(45.0).toFloat().toDouble(), (width / height.toFloat()).toDouble(), 0.1, 1000.0)
    }
}
